sdk/python/kfp/compiler/test_data/pipelines/pipeline_with_importer_and_custom_artifact_type.py

```diff
@@ -0,0 +1,78 @@
+# Copyright 2021 The Kubeflow Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Pipeline using dsl.importer."""
+
+from typing import NamedTuple
+
+from kfp import compiler
+from kfp import dsl
+from kfp.dsl import component
+from kfp.dsl import Dataset
+from kfp.dsl import importer
+from kfp.dsl import Input
+from kfp.dsl import Model
+
+
+class MyDataset(Dataset):
+    """An artifact representing a Custom Dataset."""
+    TYPE_NAME = 'custom.MyDataset'
+
+
+@component
```

Comment:
This is a bug that involves the `importer` in KFP v2, separate from use of a container component.

For instance, the following pipeline should work:
```
from kfp import dsl
from google_cloud_pipeline_components import aiplatform as gcpc_aip
from google_cloud_pipeline_components.types import artifact_types

@dsl.pipeline(name=_PIPELINE_NAME)
def pipeline():
  importer_spec = importer(
      artifact_uri=f'gs://managed-pipeline-gcpc-e2e-test/automl-tabular/model',
      artifact_class=artifact_types.UnmanagedContainerModel,
      metadata={
          'containerSpec': {
              'imageUri':
                  'us-docker.pkg.dev/vertex-ai/automl-tabular/prediction-server:prod'
          }
      })

  model_upload_with_artifact_op = gcpc_aip.ModelUploadOp(
      project=_GCP_PROJECT_ID,
      location=_GCP_REGION,
      display_name=_MODEL_DISPLAY_NAME,
      unmanaged_container_model=importer_spec.outputs['artifact'])
```

However, when compiling the above pipeline in KFP v2, the following error is encountered:
`type name UnmanagedContainerModel is different from expected: google.UnmanagedContainerModel`.

I believe this type mismatch occurs due to the logic of the output of the importer not matching the logic in component_factory._annotation_to_type_struct, hence the need for this PR.

I think for the purposes of making sure compilation passes for using a custom artifact type with the importer and an ensuing consumer component, this test would suffice.

While custom artifacts are currently unsupported in the lightweight Python component, the above example can compile and run in Managed Pipelines using KFP v1, as we're using v1 component YAML (instead of lightweight Python components). This PR is simply to make sure that the above example also runs in KFP v2 (as it should).

Reply:
Thank you for clarifying. I do see what you're saying about v1 component YAML and the purpose of the change.

I think that the compilation test is slightly misleading because it isn't executable (while the example in your comment is), but perhaps that's okay since it's not a sample test.

Do you think you could add a comment to the file explaining that the purpose is to test type compatibility at component interfaces, not to show usage of a lightweight component? And that the lightweight component would fail at runtime?

Reply:
Ok, I see your point about the pipeline being compilable but not actually executable.

I've changed the test to use a dummy YAML component to consume the output from `importer` (so that it's a pipeline meant to be both compilable and executable), and changed the name of the test to include "gcpc" (this fix works for all custom artifact types that uses `TYPE_NAME`, but perhaps it's better not to overstate the scope here).